// line-pay-sdk/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.API_LINE_PAY_URL = "https://api-pay.line.me";
exports.BETA_API_LINE_PAY_URL = "https://lgw-pay.line-apps-beta.com";
exports.SANDBOX_API_LINE_PAY_URL = "https://sandbox-api-pay.line.me";
exports.API_VERSION = "v2";
exports.CONFIG_REQUIRED_PARAMS = ["channelId", "channelSecret"];
exports.MIDDLEWARE_CONFIG_REQUIRED_PARAMS = [
    "productName",
    "amount",
    "currency",
    "orderId",
];
exports.GET_PAYMENT_DETAILS_REQUIRED_PARAMS = [];
exports.RESERVE_PAYMENT_REQUIRED_PARAMS = [
    "productName",
    "amount",
    "currency",
    "confirmUrl",
    "orderId",
];
exports.CONFIRM_PAYMENT_REQUIRED_PARAMS = [
    "transactionId",
    "amount",
    "currency",
];
exports.REFUND_PAYMENT_REQUIRED_PARAMS = ["transactionId"];
exports.GET_AUTHORIZATION_DETAILS_REQUIRED_PARAMS = [];
exports.CAPTURE_REQUIRED_PARAMS = ["transactionId", "amount", "currency"];
exports.VOID_AUTHORIZATION_REQUIRED_PARAMS = ["transactionId"];
exports.PREAPPROVED_PAYMENT_REQUIRED_PARAMS = [
    "productName",
    "amount",
    "currency",
    "orderId",
    "regKey",
];
exports.CHECK_REGKEY_STATUS_REQUIRED_PARAMS = ["regKey"];
exports.EXPIRE_REGKEY_REQUIRED_PARAMS = ["regKey"];
